src/Elmish.WPF/WpfProgram.fs

```diff
@@ -0,0 +1,137 @@
+﻿namespace Elmish.WPF
+
+open System.Windows
+open Microsoft.Extensions.Logging
+open Microsoft.Extensions.Logging.Abstractions
+open Elmish
+
+
+type WpfProgram<'model, 'msg> =
+  internal {
+    ElmishProgram: Program<unit, 'model, 'msg, Binding<'model, 'msg> list>
+    LoggerFactory: ILoggerFactory
+    /// Only log calls that take at least this many milliseconds. Default 1.
+    PerformanceLogThreshold: int
+  }
+
+
+[<RequireQualifiedAccess>]
+module WpfProgram =
+
+
+  let private fromElmishProgram program =
+    { ElmishProgram = program
+      LoggerFactory = NullLoggerFactory.Instance
+      PerformanceLogThreshold = 1 }
+
+
+  /// Creates a WpfProgram that does not use commands.
+  let mkSimple
+      (init: unit -> 'model)
+      (update: 'msg  -> 'model -> 'model)
+      (bindings: unit -> Binding<'model, 'msg> list) =
+    Program.mkSimple init update (fun _ _ -> bindings ())
+    |> fromElmishProgram
+
+
+  /// Creates a WpfProgram that uses commands
+  let mkProgram
+      (init: unit -> 'model * Cmd<'msg>)
+      (update: 'msg  -> 'model -> 'model * Cmd<'msg>)
+      (bindings: unit -> Binding<'model, 'msg> list) =
+    Program.mkProgram init update (fun _ _ -> bindings ())
+    |> fromElmishProgram
+
+
+  /// Starts the Elmish dispatch loop, setting the bindings as the DataContext
+  /// for the specified FrameworkElement. Non-blocking. This is a low-level function;
+  /// for normal usage, see runWindow and runWindowWithConfig.
+  let startElmishLoop
+      (element: FrameworkElement)
+      (program:  WpfProgram<'model, 'msg>) =
+    let mutable lastModel = None
+
+    let updateLogger = program.LoggerFactory.CreateLogger("Elmish.WPF.Update")
+    let bindingsLogger = program.LoggerFactory.CreateLogger("Elmish.WPF.Bindings")
+    let performanceLogger = program.LoggerFactory.CreateLogger("Elmish.WPF.Performance")
+
+    let setState model dispatch =
+      match lastModel with
+      | None ->
+          let bindings = Program.view program.ElmishProgram model dispatch
+          let vm = ViewModel<'model, 'msg>(model, dispatch, bindings, program.PerformanceLogThreshold, "main", bindingsLogger, performanceLogger)
+          element.DataContext <- vm
+          lastModel <- Some vm
+      | Some vm ->
+          vm.UpdateModel model
+
+    let uiDispatch (innerDispatch: Dispatch<'msg>) : Dispatch<'msg> =
+      fun msg -> element.Dispatcher.Invoke(fun () -> innerDispatch msg)
+
+    let logMsgAndModel (msg: 'msg) (model: 'model) = 
+      updateLogger.LogTrace("New message: {Message}", msg)
+      updateLogger.LogTrace("Updated state:\n{Model}", model)
```

Comment:
I think these two should be combined into one call to `LogTrace`.

Reply:
I don't categorically disagree, but may I ask why? What is the benefit of having both the message and updated state as one log statement instead of two?

A reason to keep it this way is that both the message and state may be multiline strings, so it may be easier to find what you want in logs if it's two separate entries. Also, if users want to log only messages or only state even if we use the same category, they may add logging filters to exclude one or the other. (Though if we think users will need that, we should use different log categories.)

But I could certainly be convinced the other way.

Reply:
As I said in https://github.com/elmish/Elmish.WPF/issues/105#issuecomment-524628014
> 1. [Emitting multiple log events for the same application event] makes the developer read the log history instead of providing all the relevant information in a single log event. ...
> 2. Those two log event could be written to their sink arbitrarily far apart if the application is concurrently doing other things that produce log events.

In fact, I just learned in the last week about [the features of Log and Diagnostic context in Serilog](https://benfoster.io/blog/serilog-best-practices/#log-and-diagnostic-context).  The basic idea is emitting a single log event for some application event instead of emitting more than one log event and correlating them with a correlation ID.

Reply:
> A reason to keep it this way is that both the message and state may be multiline strings, so it may be easier to find what you want in logs if it's two separate entries. Also, if users want to log only messages or only state even if we use the same category, they may add logging filters to exclude one or the other. (Though if we think users will need that, we should use different log categories.)

If you want to make it easier to ready, then add more new lines.  Maybe something like
```F#
updateLogger.LogTrace("Created new model from a message.\nMessage: {Message}\nModel: {Model}", msg, model)
```

Reply:
You've got me convinced. Just pushed a change that merges these.